operator/builtin/input/file/design.md

```diff
@@ -0,0 +1,146 @@
+2. Queueing
+    1. If the number of matched files is less than or equal to the maximum degree of concurrency, as defined by the `max_concurrent_files` setting, then no queueing occurs.
+    2. Else, queueing occurs, which means the following:
+        - Matched files are split into two sets, such that the first is small enough to respect `max_concurrent_files`, and the second contains the remaining files (called the queue).
+        - The current poll interval will begin processing the first set of files, just as if they were the only ones found during the matching phase.
+        - Subsequent poll cycles will pull matches off of the queue, until the queue is empty.
+        - The `max_concurrent_files` setting is respected at all times.
+        - TODO: Use a worker pool to process the queue as quickly as possible, rather than waiting for poll cycles to trigger batching.
+3. Opening
+    1. Each of the matched files is opened. Note:
+        - A small amount of time has passed since the file was matched.
+        - It is possible that is has been moved or deleted by this point.
+        - Only a minimum set of operations should occur between file matching and opening.
+        - If an error occurs while opening, it is logged.
+4. Fingerprinting
+    1. The first `N` bytes of each file are read. (See fingerprinting section above.)
+5. Exclusion
+    1. Empty files are closed immediately and discarded. (There is nothing to read.)
+    2. Fingerprints found in this batch are cross referenced against each other to detect duplicates. Duplicate files are closed immediately and discarded.
+        - In the vast majority of cases, this occurs during file rotation that uses the copy/truncate method. (See fingerprinting section above.)
+6. Reader Creation
+    1. Each file handle is wrapped into a `Reader` along with some metadata. (See Reader section above)
+        - During the creation of a `Reader`, the file's fingerprint is crossreferenced with previously known fingerprints.
+        - If a file's fingerprint matches one that has recently been seen, then metadata is copied over from the previous iteration of the Reader. Most importantly, the offset is accurately maintained in this way.
+        - If a file's fingerprint does not match any recently seen files, then its offset is initialized according to the `start_at` setting.
+7. Detection of "Lost" Files
+8. Consumption
+    1. Lost files are consumed. Since these files are thought to have been rotated to a location where we will not match them again, we will never see them again. 
+        - The best we can do is finish consuming their current contents.
+        - We can reasonable expect in most cases that these files are no longer being written to.
+    2. Matched files (from this poll cycle) are consumed.
+        - These file handles will be left open until the end of the next poll cycle.
```

Comment:
end of next poll cycle or end of this poll cycle? Do they stay open for the duration of the interval between cycles plus 2*duration of reading in each cycle?

Reply:
Hmm, I need to tweak this a bit. 

This is the longest a file may remain open, but it would only occur when a file is detected as "lost".

In such a case, the lifecycle of the file handle would be:

Poll Cycle N:
- File X is matched, opened, consumed.

Poll Cycle N+1:
- File X is not matched. 
- Cross referencing with files from Poll Cycle N shows that File X has been "lost".
- Remainder of File X is consumed.
- File X is closed.

The typical file handle lifecycle would be a bit shorter:

Poll Cycle N:
- File X is matched, opened, consumed.

Poll Cycle N+1:
- File X is matched again.
- Cross referencing with files from Poll Cycle N shows we now have duplicate handles to File X.
- Close old handle to File X.